sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -340,13 +358,26 @@ trait ExposesMetadataColumns extends LogicalPlan {
       val resolve = conf.resolver
       val outputNames = outputSet.map(_.name)
 
-      def isOutputColumn(col: AttributeReference): Boolean = {
-        outputNames.exists(name => resolve(col.name, name))
+      // Generate a unique name by prepending underscores.
+      @scala.annotation.tailrec
+      def makeUnique(name: String): String = name match {
+        case name if outputNames.exists(resolve(_, name)) => makeUnique(s"_$name")
+        case name => name
+      }
+
+      // Rename metadata struct columns whose names conflict with output columns.
```

Comment:
Name is not the id of a column, e.g. people can do `SELECT 1 a, 2 a`. I don't quite understand why we need to rename here. What we need is to resolve `_metadata` to the actual metadata `AttributeReference`. At the end Spark will bind `AttributeReference` correctly by looking at `exprId`.

Reply:
The immediate issue is that node types implementing `ExposesMetadataColumns` all use its `metadataOutputWithOutConflicts` method to ensure that names metadata columns in `metadataOutput` do not overlap with names of non-metadata columns in `output`. Today's code simply removes conflicting metadata attributes; this PR renames them instead. 

We could potentially explore removing that method entirely, but I don't know what side effects that might have?

Reply:
A quick local test of the change, with this query over a table with conflicting user schema:
```scala
df.select("_metadata")
  .withColumn("file_name", df.metadataColumn("_metadata").getField("file_name"))
```

... produces the expected logical plan (names conflict but `exprId` differs):
```
Project [_metadata#20, _metadata#24.file_name AS file_name#26]
+- Project [_metadata#20, _metadata#24]
   +- Relation [name#18,age#19,_METADATA#20,_metadata#24] parquet
```

... but query optimization fails with: 
```
[info]   java.lang.IllegalArgumentException: file_name does not exist. Available:
[info]   at org.apache.spark.sql.types.StructType.$anonfun$fieldIndex$1(StructType.scala:313)
[info]   at scala.collection.immutable.Map$EmptyMap$.getOrElse(Map.scala:110)
[info]   at org.apache.spark.sql.types.StructType.fieldIndex(StructType.scala:312)
  ...
[info]   at org.apache.spark.sql.catalyst.expressions.ProjectionOverSchema.getProjection(ProjectionOverSchema.scala:71)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.$anonfun$buildNewProjection$3(SchemaPruning.scala:154)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.buildNewProjection(SchemaPruning.scala:154)
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.org$apache$spark$sql$execution$datasources$SchemaPruning$$prunePhysicalColumns(SchemaPruning.scala:99)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.apply(SchemaPruning.scala:41)
  ...
[info]   at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:211)
  ...
[info]   at org.apache.spark.sql.execution.QueryExecution.optimizedPlan$lzycompute(QueryExecution.scala:139)
  ...
[info]   at org.apache.spark.sql.QueryTest.checkAnswer(QueryTest.scala:148)
[info]   at org.apache.spark.sql.execution.datasources.FileMetadataStructSuite.$anonfun$new$13(FileMetadataStructSuite.scala:255)
```
So it appears that at least one optimizer rule fails to honor expression ids, and assumes the column names are unique?

Reply:
@viirya this seems like a bug in nested column pruning

Reply:
Looks like `getProjection` wrongly matches attribute. It tries to get pruned datatype from `schema` by looking with field name. `StructType` maintains the mapping between (field name -> StructField), not (expr id -> StructField). If there are duplicate columns, it possibly picks incorrect attribute.

Reply:
> If there are duplicate columns, it possibly picks incorrect attribute.

Is it possible to keep both in the requested schema?  

cc @cashmand

Reply:
I don't think there's anything in `StructType` that could track expr id (and IMO it shouldn't be trying to do so -- different levels of abstraction). 

It really looks like `StructType` doesn't allow duplicate names. The class is full of code that assumes uniqueness, often leveraging these private members: 
```scala
  private lazy val fieldNamesSet: Set[String] = fieldNames.toSet
  private lazy val nameToField: Map[String, StructField] = fields.map(f => f.name -> f).toMap
  private lazy val nameToIndex: Map[String, Int] = Utils.toMapWithIndex(fieldNames)
```
The `fieldNamesSet` goes all the way back to 2015...